design/004-external-data-handling/004-external-data-handling.md

```diff
@@ -0,0 +1,232 @@
+recorded.
+
+#### Upcoming
+
+We have plans to include other external data providers, such as NFS shares,
+openBIS etc. From the user's perspective, they all will generally work in the
+same way as S3; data will be mounted to a local directory outside the repository
+and made accessible as network-mounted storage.
+
+## Design Detail
+
+Generically, we should strive to remove the separation between "hosted" Renku
+(i.e. what I can do if I run an interactive session on renkulab.io) and
+"offline" Renku (what I can do with the CLI anywhere).
+
+Ideally we could once again follow the plug-in paradigm in the CLI, like we
+do for other functionality.
+
+### Provider interface
+
+tbd
+
+### Commands
+
+#### Adding external data:
+
+```
+renku dataset add s3://server/path/to/bucket
+Access Key: 1234
+Secret Key: abcd
+Store credentials? (y/N): y
+Credentials stored in ~/.renku/renku.ini
+```
+
+`s3://` is used here but we should also support at least `nfs://`. Renku should
+ask for credentials in a way that is appropriate for the type of resource that
+is being added.
+
+#### Getting information about the dataset:
+
+```
+renku dataset show very-large-data
+
+Name: very-large-data
+Created: 2022-03-30 15:55:59+02:00
+Creator(s): Rok Roškar <rok.roskar@sdsc.ethz.ch>
+Title: very-large-data
+Description: Data that is very large. Indeed.
+External Sources:
+    - type: s3
+      source: s3://server/path/to/bucket
+      location: None
+```
+
+#### Staging external data:
+
+External data has to somehow be brought into the system where Renku is running.
+This seems to be a "staging" operation, but we could also call it something
+else. "Mounting" is another option but seems too linux-specific.
+
+```
+renku dataset stage very-large-data
+```
+
+This command uses Fuse when necessary to mount data locally. Once it's staged,
+it can be found locally and used in workflows:
+
+```
+renku dataset show very-large-data
+
+Name: very-large-data
+Created: 2022-03-30 15:55:59+02:00
+Creator(s): Rok Roškar <rok.roskar@sdsc.ethz.ch>
+Title: very-large-data
+Description: Data that is very large. Indeed.
+External Sources:
+    - type: s3
+      source: s3://server/path/to/<bucket>
+      location: .renku/mounts/<bucket>
+```
+
+The data is accessible under `data/very-large-data/<bucket>` in the project's directory.
+
+If the data is already mounted by some other means, `dataset stage` can be used
+to let Renku know about it:
+
+```
+renku dataset stage --existing s3://server/path/to/<bucket>:/mounts/bucket
+```
+
+In managed interactive sessions this should happen automatically.
+
+#### Streaming external data
+
+The renku-python API should offer some support for more easily accessing remote
+data, especially for sources from which it is common for data to be processed as
+a stream. Ideally, Renku could lazily stage the external data if access to the
+resource is requested.
+
+## Drawbacks
+
+* Relying on fuse is tricky because it requires elevated privileges on all
```

Comment:
@olevski do you have any insights on this?

Reply:
so locally you don't need escalated privileges, as long as you have access to `/dev/fuse` and in an interactive session, we could do it in an escalated sidecare with a microservice.

Reply:
I do not have really much experience with using fuse on personal laptop. I thought that you do need sudo but I guess that is not always the case. I just did `sshfs` on my mac and that did not require sudo.

Reply:
IMO this feels a really big issue to tackle. Can we split it up in two pieces:
1. How to track provenance of external data
2. How to mount nfs/s3/etc with the cli

I think it may be a bit easier to think of these separately.

Also trying to get all this mounting to work on 2/3 operating systems seems very intimidating.

Reply:
I got some background and more info from @Panaetius on this. And I feel better now :)

Reply:
@Panaetius I'd be curious to know what this comforting information was... :)

Reply:
> I got some background and more info from @Panaetius on this. And I feel better now :)

@olevski I have the same feeling- what did you learn that made you feel better?

Reply:
I'd also like to know!

Reply:
Lol it was mostly that we would try to implement the cli and notebook service components separately